src/support/topological_sort.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_support_topological_sort_h
+#define wasm_support_topological_sort_h
+
+namespace wasm {
+
+template<typename T, typename Subtype> struct TopologicalSort {
+private:
+  // The DFS work list.
+  std::vector<T> workStack;
+
+  // Remember which items we have finished so we don't visit them again.
+  std::unordered_set<T> finished;
+
+  // Should be overridden by `Subtype`.
+  void pushPredecessors(T item) {
+    static_assert(&TopologicalSort<T, Subtype>::pushPredecessors !=
+                    &Subtype::pushPredecessors,
+                  "TopologicalSort subclass must implement `pushPredecessors`");
+    WASM_UNREACHABLE("Derived class must implement pushPredecessors");
+  }
+
+  // Pop until the stack is empty or it has an unfinished item on top.
+  void pop() {
```

Comment:
About the name, this doesn't return an item, which popping usually does (or should). It is also not a single pop. Maybe `finishTop`? (as it finishes the top item, then does the logical work resulting from that)